src/Workspaces/Remote/Core/Services/AssetSource.cs

```diff
@@ -13,21 +13,14 @@ namespace Microsoft.CodeAnalysis.Remote
     internal abstract class AssetSource
     {
         private readonly AssetStorage _assetStorage;
-        private readonly int _scopeId;
 
-        protected AssetSource(AssetStorage assetStorage, int scopeId)
+        protected AssetSource(AssetStorage assetStorage)
         {
             _assetStorage = assetStorage;
-            _scopeId = scopeId;
 
-            _assetStorage.RegisterAssetSource(_scopeId, this);
+            _assetStorage.SetAssetSource(this);
```

Comment:
:memo: This back-reference leaves me feeling uneasy. It makes it hard to understand the pre- and post-conditions for using this type.